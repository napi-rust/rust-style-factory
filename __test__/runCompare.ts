import { readFileSync } from 'fs-extra';
import styleFactory from 'style-factory';
import { styleFactory as styleFactoryRust } from '../index';
import { expect } from 'vitest';
import path, { dirname } from 'node:path';
import { writeFileSync } from 'node:fs';
import { transform } from 'lightningcss';
import { requireModuleCode } from 'require-module-exports';
import prettier from 'prettier';

function minifyCss(css: string) {
  return new Promise<string>((resolve) => {
    resolve(
      transform({
        filename: 'a.css',
        code: Buffer.from(css, 'utf8'),
        minify: true,
      }).code.toString(),
    );
  });
}

async function formatCss(code: string) {
  try {
    return await prettier.format(code, {
      parser: 'css',
      tabWidth: 2,
      singleQuote: true,
      trailingComma: 'all',
      semi: true,
    });
  } catch (e) {
    return code;
  }
}

async function writeFile(file: string, code: string) {
  const ext = path.extname(file);
  const isCss = ext === '.css';
  const formatCode = await prettier.format(code, {
    parser: isCss ? 'css' : 'babel',
    tabWidth: 2,
    singleQuote: true,
    trailingComma: 'all',
    semi: true,
  });
  writeFileSync(file, formatCode);
}

export const runCompare = async (file: string) => {
  const context = readFileSync(file);
  const dir = dirname(file);

  const code = await minifyCss(context.toString());

  await writeFile(file.replace('.css', '.min.css'), code);

  const jsCode = styleFactory(code, {
    transformTag: (t) => {
      if (t === '\\*') {
        return 'unsupport-star';
      }
      if (t === 'page') {
        return '[meta\\\\:tag=page]';
      }
      return `[meta\\\\:tag=${t}]`;
    },
  });
  const rustCode = styleFactoryRust(code);

  await writeFile(path.join(dir, 'js.js'), jsCode);
  await writeFile(path.join(dir, 'rs.js'), rustCode);

  const jsFn = requireModuleCode(jsCode).default;
  const rustFn = requireModuleCode(rustCode).default;

  const options = {
    prefix: 'sf-',
    rpx: (value: number) => value / 2,
    tag: (tag: string) => {
      return tag;
    },
    host: 'component-666',
  };
  const jsResult = jsFn(options);
  const rustResult = rustFn(options);

  const jsCss = await minifyCss(jsResult);
  const rustCss = await minifyCss(rustResult);

  expect(jsCss).toMatchSnapshot('js');
  expect(rustCss).toMatchSnapshot('rs');

  await writeFile(path.join(dir, 'js.css'), jsCss);
  await writeFile(path.join(dir, 'rust.css'), rustCss);

  expect(jsCss.length).toEqual(rustCss.length);
  expect(jsCss).toEqual(rustCss);
};
