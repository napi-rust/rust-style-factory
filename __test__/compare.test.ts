import fs from 'node:fs';
import path from 'node:path';
import { describe, expect, it, vi } from 'vitest';
import { runCompare } from './runCompare';

vi.setConfig({
  testTimeout: 20_000,
});

const fixtures = path.join(__dirname, '../__fixtures__');

const cases = fs
  .readdirSync(fixtures)
  .filter((f) => {
    return fs.statSync(path.join(fixtures, f)).isDirectory();
  })
  .filter((f) => !f.startsWith('.'))
  .filter((f) => {
    if (process.env.filter) {
      return f.includes(process.env.filter);
    }
    return true;
  });

const ignore = ['import', 'import2', 'large'];

for (const f of cases) {
  if (ignore.includes(f)) {
    continue;
  }
  const cssFile = path.join(fixtures, f, 'index.css');
  describe(`runCompile ${f}`, async () => {
    it('style-factory', async () => {
      runCompare(cssFile);
    });
  });
}
