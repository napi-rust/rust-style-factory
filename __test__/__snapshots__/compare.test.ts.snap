// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runCompile all > style-factory > js 1`] = `".sf-foo{color:red;height:50px}#id,.sf-class,[is=component-666]{color:red}.sf-foo{color:green}.sf-a .sf-b,.sf-a>.sf-b,.sf-a~.sf-b,.sf-a+.sf-b,.sf-a:after{color:red}:root{--main-bg-color:#fefefe;--main-text-color:#363636}[theme=dark]:root{--main-bg-color:#2f3a44;--main-text-color:#c5c5c5}"`;

exports[`runCompile all > style-factory > rs 1`] = `".sf-foo{color:red;height:50px}#id,.sf-class,[is=component-666]{color:red}.sf-foo{color:green}.sf-a .sf-b,.sf-a>.sf-b,.sf-a~.sf-b,.sf-a+.sf-b,.sf-a:after{color:red}:root{--main-bg-color:#fefefe;--main-text-color:#363636}:root[theme=dark]{--main-bg-color:#2f3a44;--main-text-color:#c5c5c5}"`;

exports[`runCompile calc > style-factory > js 1`] = `".sf-h1{height:15px}.sf-h2{width:35px}"`;

exports[`runCompile calc > style-factory > rs 1`] = `".sf-h1{height:calc(50rpx - 20rpx)}.sf-h2{width:calc(50rpx - -20rpx)}"`;

exports[`runCompile comment > style-factory > js 1`] = `".sf-foo{color:red;height:50px}"`;

exports[`runCompile comment > style-factory > rs 1`] = `".sf-foo{color:red;height:50px}"`;

exports[`runCompile empty > style-factory > js 1`] = `""`;

exports[`runCompile empty > style-factory > rs 1`] = `""`;

exports[`runCompile font-face > style-factory > js 1`] = `"@font-face{font-family:iconfont;src:url(/components/foo/iconfont.woff2?t=1642945972391)format("woff2")}"`;

exports[`runCompile font-face > style-factory > rs 1`] = `"@font-face{font-family:iconfont;src:url(/components/foo/iconfont.woff2?t=1642945972391)format("woff2")}"`;

exports[`runCompile host > style-factory > js 1`] = `"[is=component-666]{color:red;width:50px;font-size:10px}.sf-a [is=component-666]{height:50px}"`;

exports[`runCompile host > style-factory > rs 1`] = `".sf-a [is=component-666]{height:50px}"`;

exports[`runCompile important > style-factory > js 1`] = `".sf-a{position:absolute!important}"`;

exports[`runCompile important > style-factory > rs 1`] = `".sf-a{position:absolute!important}"`;

exports[`runCompile keyframes > style-factory > js 1`] = `"@keyframes slidein{0%{transform:translate(0%)}to{transform:translate(100%)}}"`;

exports[`runCompile keyframes > style-factory > rs 1`] = `"@keyframes slidein{0%{transform:translate(0%)}to{transform:translate(100%)}}"`;

exports[`runCompile keyframes-2 > style-factory > js 1`] = `"@keyframes anim-show{to{opacity:1}}@keyframes anim-hide{to{opacity:0}}"`;

exports[`runCompile keyframes-2 > style-factory > rs 1`] = `"@keyframes anim-show{to{opacity:1}}@keyframes anim-hide{to{opacity:0}}"`;

exports[`runCompile media > style-factory > js 1`] = `"@media screen and (min-width:900px){.sf-article{padding:1rem 3rem}}"`;

exports[`runCompile media > style-factory > rs 1`] = `"@media screen and (min-width:900px){.sf-article{padding:1rem 3rem}}"`;

exports[`runCompile media2 > style-factory > js 1`] = `"@supports (display:flex){@media screen and (min-width:900px){.sf-article{display:flex}}}"`;

exports[`runCompile media2 > style-factory > rs 1`] = `"@supports (display:flex){@media screen and (min-width:900px){.sf-article{display:flex}}}"`;

exports[`runCompile not > style-factory > js 1`] = `".sf-a:not(.sf-b){color:red}"`;

exports[`runCompile not > style-factory > rs 1`] = `".sf-a:not(.sf-b){color:red}"`;

exports[`runCompile nth-child > style-factory > js 1`] = `"ty-button:nth-child(2){color:red}"`;

exports[`runCompile nth-child > style-factory > rs 1`] = `"[meta\\:tag=ty-button]:nth-child(2){color:red}"`;

exports[`runCompile nth-child2 > style-factory > js 1`] = `".sf-ty-scroll-view-loading-black .sf-ty-scroll-view-dot:first-child,.sf-ty-scroll-view-loading-white .sf-ty-scroll-view-dot:first-child{animation:.9s infinite ty-scroll-view-keyframes}.sf-ty-scroll-view-loading-black .sf-ty-scroll-view-dot:nth-child(2),.sf-ty-scroll-view-loading-white .sf-ty-scroll-view-dot:nth-child(2){animation:.9s .3s infinite ty-scroll-view-keyframes}.sf-ty-scroll-view-loading-black .sf-ty-scroll-view-dot:nth-child(3),.sf-ty-scroll-view-loading-white .sf-ty-scroll-view-dot:nth-child(3){animation:.9s .6s infinite ty-scroll-view-keyframes}"`;

exports[`runCompile nth-child2 > style-factory > rs 1`] = `".sf-ty-scroll-view-loading-black .sf-ty-scroll-view-dot:first-child,.sf-ty-scroll-view-loading-white .sf-ty-scroll-view-dot:first-child{animation:.9s infinite ty-scroll-view-keyframes}.sf-ty-scroll-view-loading-black .sf-ty-scroll-view-dot:nth-child(2),.sf-ty-scroll-view-loading-white .sf-ty-scroll-view-dot:nth-child(2){animation:.9s .3s infinite ty-scroll-view-keyframes}.sf-ty-scroll-view-loading-black .sf-ty-scroll-view-dot:nth-child(3),.sf-ty-scroll-view-loading-white .sf-ty-scroll-view-dot:nth-child(3){animation:.9s .6s infinite ty-scroll-view-keyframes}"`;

exports[`runCompile star > style-factory > js 1`] = `"*{padding:0}page,.sf-abc *{color:red}"`;

exports[`runCompile star > style-factory > rs 1`] = `"unsupport-star{padding:0}[meta\\:tag=page],.sf-abc unsupport-star{color:red}"`;

exports[`runCompile tags > style-factory > js 1`] = `"view+button,web-view,button{color:#363636}"`;

exports[`runCompile tags > style-factory > rs 1`] = `"[meta\\:tag=view]+[meta\\:tag=button],[meta\\:tag=web-view],[meta\\:tag=button]{color:#363636}"`;

exports[`runCompile url > style-factory > js 1`] = `".sf-h1:not(.sf-foo){background-image:url(abc\\).sbc)}"`;

exports[`runCompile url > style-factory > rs 1`] = `".sf-h1:not(.sf-foo){background-image:url(abc\\).sbc)}"`;
